src/curl_jolokia.c

```diff
@@ -0,0 +1,1100 @@
+  int attribute_pool_used;
+
+  const char *json_key;
+  yajl_len_t json_key_len;
+
+  cjo_expect_token expect;
+
+  _Bool verify_peer;
+  _Bool verify_host;
+  char *cacert;
+  struct curl_slist *headers;
+  char *post_body;
+  size_t post_body_len;
+  _Bool expect_escapes;
+  cdtime_t interval;
+  int timeout;
+
+  CURL *curl;
+  char curl_errbuf[CURL_ERROR_SIZE];
+  const unsigned char *start, *end;
+  cjo_membuffer_t replybuffer;
+  cjo_membuffer_t itembuffer;
+
+  size_t poolsize;
+  char *pool;
+  const char *poolmax;
+  char *pool_ptr;
+
+  yajl_handle yajl;
+  llist_t *bean_configs;
+  int max_attribute_count;
+  int max_value_len;
+};
+typedef struct cjo_s cjo_t; /* }}} */
+
+static void cjo_init_buffer(cjo_membuffer_t *buf, size_t initial_size) {
+  buf->buffer = malloc(initial_size);
+  buf->used = 0;
+  buf->size = initial_size;
+}
+
+static void cjo_append_buffer(cjo_membuffer_t *buf, const char *append,
+                              size_t appendlen) {
+  if (buf->used + appendlen + 1 > buf->size) {
+    char *newbuf;
+    size_t newsize;
+
+    newsize = buf->size * 2;
+
+    newbuf = malloc(newsize);
+    memcpy(newbuf, buf->buffer, buf->used);
+    free(buf->buffer);
+    buf->buffer = newbuf;
+    buf->size = newsize;
+  }
+  memcpy(buf->buffer + buf->used, append, appendlen);
+  buf->used += appendlen;
+}
+
+static void cjo_release_buffercontent(cjo_membuffer_t *buf) {
+  sfree(buf->buffer);
+  memset(buf, 0, sizeof(cjo_membuffer_t));
+}
+
+static void cjo_remember_value(cjo_membuffer_t *ShouldBeInHere,
+                               cjo_membuffer_t *BufferHereIfNot,
+                               const char *ptr, size_t len,
+                               const char **SetThisOne,
+			       yajl_len_t *SetThisLen) {
+  if (!((ptr > ShouldBeInHere->buffer) &&
+        (ptr < ShouldBeInHere->buffer + ShouldBeInHere->used))) {
+    char *tptr;
+    tptr = BufferHereIfNot->buffer + BufferHereIfNot->used;
+    cjo_append_buffer(BufferHereIfNot, ptr, len);
+    ptr = tptr;
+  }
+
+  *SetThisOne = ptr;
+  *SetThisLen = (yajl_len_t) len;
+}
+
+/*
+ * llist lookup methods
+ */
+static int cjo_cfg_compare_attribute(llentry_t *match, void *vdb) {
+  cjo_t *db = (cjo_t *)vdb;
+  cjo_attribute_t *pcfg = (cjo_attribute_t *)match->value;
+
+  if (pcfg->attribute_match_len != db->curr_attribute->json_name_len)
+    return -1;
+  else
+    return strncmp(pcfg->attribute_match, db->curr_attribute->json_name,
+                   db->curr_attribute->json_name_len);
+}
+static int cjo_cfg_compare_bean(llentry_t *match, void *vdb) {
+  cjo_t *db = (cjo_t *)vdb;
+  cjo_bean_t *pbeancfg = (cjo_bean_t *)match->value;
+
+  if (pbeancfg->mbean_match_len != db->json_key_len)
+    return -1;
+  return strncmp(pbeancfg->mbean_match, db->json_key, db->json_key_len);
+}
+
+/*
+ * Aligning data with Configs & Sending it
+ */
+
+static int cjo_get_type(cjo_attribute_t *attribute) {
+  const data_set_t *ds;
+
+  assert(attribute->type != NULL);
```

Comment:
Please don't do argument checking with `assert()`. Either

1.  add a comment along the lines of `/* ensured in cjo_somefunc */`, or
2.  change to something like `if (attr->type == NULL) { return EINVAL; };`